### PR TITLE
updating cell browser (first try)

### PR DESCRIPTION
**Pull request by maximilianh**
Hi, this is an operation that I've never done before. I forked your repo, cherry picked the commits from your PR, then modified cellbrowser.R, commited the result and made a PR from that. 

1) How can I test the result now? 

2) the function tries to support seurat2 and seurat3. Is this acceptable?

3) a part of the new function does not work on Windows, but it's not an essential part. OK?

4) I don't understand why I'm not getting a diff in the Github interface... all lines seem to be new... hmm...

### CODE REVIEW

File: R/cellbrowser.R [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+  matrix.slot = "counts",
+  markers.n = 100,
+  skip.markers = FALSE,
+  skip.expr.matrix = FALSE,
+  skip.metadata = FALSE,
+  skip.reductions = FALSE
+) {
+  if (!requireNamespace("Seurat", quietly = TRUE)) {
+    stop("This script requires that Seurat (V2 or V3) is installed")
+  }
```
Comment (mojaveazure) [inline R/cellbrowser.R:168]: At least in SeuratWrappers, Seurat is assumed to be installed; is this check necessary?

File: R/cellbrowser.R [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+  message("Seurat Version installed: ", packageVersion(pkg = "Seurat"))
```
Comment (mojaveazure) [inline R/cellbrowser.R:169]: Functions not included in base or SeuratWrappers must be imported; `packageVersion` is part of `utils`

File: R/cellbrowser.R [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+  object.version <- as.character(x = slot(object = object, name = 'version'))
+  message("Object was created with Seurat version ", object.version)
+  if (substr(x = object.version, start = 1, stop = 1) != '2' && substr(x = object.version, start = 1, stop = 1) != '3') {
+    stop("can only process Seurat2 or Seurat3 objects, version of rds is ", object.version)
+  }
+  if (substr(x = object.version, start = 1, stop = 1) != substr(x = packageVersion(pkg = "Seurat"), start = 1, stop = 1)) {
+    stop("The installed version of Seurat is different from Seurat object loaded. You have to down- or upgrade your installed Seurat version, see the Seurat documentation")
+  }
```
Comment (mojaveazure) [inline R/cellbrowser.R:177]: Could a lot of this be done with math on `numeric_version` objects? Also, SeuratWrappers only works on Seurat v3.0.1 or higher; you cannot have SeuratWrappers and Seurat v2 installed at the same time

File: R/cellbrowser.R [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+  embeddings <- reductions # Seurat prefers the word reductions
```
Comment (mojaveazure) [inline R/cellbrowser.R:178]: What's the purpose of this line? The cell embeddings are a specific part of the dimensional reduction; however, throughout Seurat and SeuratWrappers, we use "reductions" to refer to a dimensional reduction that we may take parts of (such as the cell embeddings) later

File: R/cellbrowser.R [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+      counts <- object@data
+    } else {
+      stop("matrix.slot can only be one of: counts, scale.data, data")
+    }
+    genes <- rownames(x = object@data)
+    dr <- object@dr
+  } else {
+    # Seurat 3 functions
+    idents <- Idents(object = object)
+    meta <- object[[]]
```
Comment (mojaveazure) [inline R/cellbrowser.R:201]: I changed this from `object@meta.data` to match the Seurat v3 way of pulling the metadata data frame

File: R/cellbrowser.R [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+    else if (matrix.slot == "data") {
+      counts <- object@data
+    } else {
+      stop("matrix.slot can only be one of: counts, scale.data, data")
+    }
+    genes <- rownames(x = object@data)
+    dr <- object@dr
+  } else {
+    # Seurat 3 functions
+    idents <- Idents(object = object)
```
Comment (mojaveazure) [inline R/cellbrowser.R:200]: Functions from Seurat also need to be declared with `@importFrom`

File: R/cellbrowser.R [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+    meta <- object[[]]
+    cellOrder <- colnames(x = object)
+    counts <- findMatrix(object = object, matrix.slot = matrix.slot)
+    if (dim(x = counts)[1] == 0) {
```
Comment (mojaveazure) [inline R/cellbrowser.R:204]: See [`IsMatrixEmpty`](https://github.com/satijalab/seurat-wrappers/blob/master/R/internal.R#L165:L177)

File: R/cellbrowser.R [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+    counts <- GetAssayData(object = object, slot = "counts")
+  } else if (matrix.slot == "scale.data") {
+    counts <- GetAssayData(object = object, slot="scale.data")
+  }
+  else if (matrix.slot=="data") {
+    counts <- GetAssayData(object = object)
+  } else {
+    stop("matrix.slot can only be one of: counts, scale.data, data")
+  }
+}
```
Comment (mojaveazure) [inline R/cellbrowser.R:81]: What's the difference between `findMatrix` and `GetAssayData`? Also, it doesn't like `findMatrix` returns anything

File: R/cellbrowser.R [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+      genes <- rownames(counts)
+      if (dim(x = counts)[1] == 0) {
+        stop(
+          "Could not find an expression matrix",
+          "Please select the correct slot where the matrix is stored, possible ",
+          "values are 'counts', 'scale.data' or 'data'. To select a slot, ",
+          "use the option 'matrix.slot' from R or the cbImportSeurat option -s from the command line."
+        )
+      }
+    }
```
Comment (mojaveazure) [inline R/cellbrowser.R:219]: I don't know what this section will accomplish. You're already using the default assay (I didn't see an assay switch for the first attempt to find expression data), so I think you're just trying to pull the same data again. Each Seurat v3 assay must have the `data` slot filled; you could try falling back to that if the first data slot was empty?

File: R/cellbrowser.R [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+    else {
+      genes <- rownames(x = object)
+    }
+    dr <- object@reductions
```
Comment (mojaveazure) [inline R/cellbrowser.R:223]: Typically, we as that you use the accessors or `slot` to pull data from a Seurat v3 object instead of `@`

File: R/cellbrowser.R [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+  }
+  if (dataset.name == "SeuratProject") {
+    warning("Using default project name means that you may overwrite project with the same name in the cellbrowser html output folder")
+  }
+  enum.fields <- c()
+  # Export expression matrix
+  if (!skip.expr.matrix) {
+    if (use.mtx) {
+      require(Matrix)
+      require(R.utils)
```
Comment (mojaveazure) [inline R/cellbrowser.R:251]: CRAN dependencies should be declared with `@importFrom` and listed in the `Imports:` section of DESCRIPTION, not called with `require`

File: R/cellbrowser.R [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+    if (length(levels(idents)) > 1) {
+      markers.helper <- function(x) {
+        partition <- markers[x,]
+        ord <- order(partition$p_val_adj < 0.05, -partition$avg_logFC)
+        res <- x[ord]
+        naCount <- max(0, length(x) - markers.n)
+        res <- c(res[1:markers.n], rep(NA, naCount))
+        return(res)
+      }
+      if (.hasSlot(object, "misc") && !is.null(x = object@misc["markers"][[1]])) {
```
Comment (mojaveazure) [inline R/cellbrowser.R:374]: See the `Misc` function from Seurat

File: R/cellbrowser.R [hunk 13]
```diff
@@ -0,0 +1,499 @@
...
+      fnames <- append(x = fnames, values = sliceFname)
+      startIdx <- startIdx + sliceSize
+    }
+    message("Concatenating chunks")
+    system(command = paste(
+      "cat",
+      paste(fnames, collapse = " "), "| gzip >",
+      outFname,
+      sep = " "
+    ))
```
Comment (mojaveazure) [inline R/cellbrowser.R:59]: An OS check should be implemented in this function to prevent running on Windows

File: R/cellbrowser.R [hunk 14]
```diff
@@ -0,0 +1,499 @@
...
+  if (substr(x = object.version, start = 1, stop = 1) != '2' && substr(x = object.version, start = 1, stop = 1) != '3') {
+    stop("can only process Seurat2 or Seurat3 objects, version of rds is ", object.version)
+  }
+  if (substr(x = object.version, start = 1, stop = 1) != substr(x = packageVersion(pkg = "Seurat"), start = 1, stop = 1)) {
+    stop("The installed version of Seurat is different from Seurat object loaded. You have to down- or upgrade your installed Seurat version, see the Seurat documentation")
+  }
+  embeddings <- reductions # Seurat prefers the word reductions
+  # compatibility layer for Seurat 2 vs 3
+  # see https://satijalab.org/seurat/essential_commands.html
+  if (substr(x = object.version, start = 1, stop = 1) == '2') {
```
Comment (mojaveazure) [inline R/cellbrowser.R:181]: This could be replaced with `if (inherits(x = object, what = 'seurat'))` as Seurat v2 objects are called `seurat` while v3 objects are called `Seurat`; you could also define separate S3 or S4 methods to dispatch differently between the two classes

File: R/cellbrowser.R (hunk 1 above)
Comment (maximilianh) [reply R/cellbrowser.R:168]: Not necessary here, but could I leave it in? It has happened to me and colleagues a few times that we got very strange error messages in pipelines and this just confirms something which should be the case anyways and for the unlikely case that it happens, has a nice error message.

File: R/cellbrowser.R (hunk 2 above)
Comment (maximilianh) [reply R/cellbrowser.R:169]: done.

File: R/cellbrowser.R (hunk 3 above)
Comment (maximilianh) [reply R/cellbrowser.R:177]: 1) What stops me from installing seuratWrappers when I have Seurat2 installed? 

2) nice! I rewrote this with "numeric_version" now.

File: R/cellbrowser.R (hunk 4 above)
Comment (maximilianh) [reply R/cellbrowser.R:178]: Sorry, I never understood the difference and I don't understand your explanation. I changed the variable now to "reducNames", hope this is better.

File: R/cellbrowser.R (hunk 5 above)
Comment (maximilianh) [reply R/cellbrowser.R:201]: OK. (doesn't seem more readable to me than @meta.data, but OK)

File: R/cellbrowser.R (hunk 12 above)
Comment (maximilianh) [reply R/cellbrowser.R:374]: Does that work with Seurat2? this line will work with Seurat2.

File: R/cellbrowser.R (hunk 13 above)
Comment (maximilianh) [reply R/cellbrowser.R:59]: done. did the check in the calling function.

File: R/cellbrowser.R (hunk 14 above)
Comment (maximilianh) [reply R/cellbrowser.R:181]: done

File: R/cellbrowser.R (hunk 11 above)
Comment (maximilianh) [reply R/cellbrowser.R:251]: done.

**Comment by maximilianh** [comment]
> Yes, that's fine. Note, you won't get access to Seurat v2 functions, only the object. 
Sorry I dont understand, what does this mean? Aren't objects created in R with functions?

> SeuratWrappers only 
> works with Seurat v3 installed, but v3 provides both the v2 (as seurat) and v3 (as Seurat) classes, so you can easily 
> test or define methods for each.
Sorry, I don't understand. Can I do a "require(seurat)" when Seurat3 is installed and get all Seurat2 classes and functions?
My current is simply a big if-then-else, as you saw. Won't this work just fine?

**Comment by mojaveazure** [comment]
> Yes, that's fine. Note, you won't get access to Seurat v2 functions, only the object.
> Sorry I dont understand, what does this mean? Aren't objects created in R with functions?

It means that you access to the class definition, but no setters/accessors as defined in Seurat v2.X. Both the v2 and v3 objects were defined as S4 classes (similar to Python classes) and are completely independent from one another. The v2 object (class `seurat`) does neither inherits nor is a superclass of the v3 object (class `Seurat`). This means that we need different code to work with v2 `seurat` and v3 `Seurat` objects. Seurat v3.X defines setters and accessors that work exclusively on the Seurat v3 `Seurat` object, and implements _no_ methods for pulling data from, adding data to, or working with v2 `seurat` objects. The function to create an S4 object (`methods::new`) is able to create a v2 `seurat` object because the class definition is present, but no other functionality is defined. If we didn't provide the v2 `seurat` class definition, then there would be no way to load, create, or work with v2 `seurat` objects in Seurat v3.X. The only Seurat v3.X function that accepts a v2 `seurat` object is `UpdateSeuratObject`, where we convert a v2 `seurat` object to a v3 `Seurat` object.

TLDR: the v2 `seurat` and v3 `Seurat` objects are fundamentally different and not interchangeable. Seurat v3.X defines no code for working with v2 `seurat` objects and SeuratWrappers works exclusively with Seurat v3.X

> Sorry, I don't understand. Can I do a "require(seurat)" when Seurat3 is installed and get all Seurat2 classes and functions?
My current solution is simply a big if-then-else, as you saw. Won't this work just fine?

No, you get the v2 `seurat` object and no other Seurat v2.X function. You will not be able to run `FindMarkers` on a v2 `seurat` object with Seurat v3.X loaded since Seurat v3.X does not define any functionality for v2 `seurat` objects. The if/else for pulling data will work, but that's about the only thing you can do with a v2 `seurat` object in SeuratWrappers without defining custom functionality.

File: R/cellbrowser.R (hunk 5 above)
Comment (mojaveazure) [reply R/cellbrowser.R:241]: It is a little obscure, but we encourage the use of accessors/setters over directly accessing slots. This gives us the freedom to modify the object structure as needed while preserving data access code for the user/developer. Basically, if we decide that we need to move or restructure the cell-level metadata at the object level, we can as long as we update the `[[` method to return a data frame the way our users are used to.

For more details about getters and setters in Seurat v3, please see our [command cheatsheet](https://satijalab.org/seurat/essential_commands.html)

File: R/cellbrowser.R (hunk 3 above)
Comment (mojaveazure) [reply R/cellbrowser.R]: [This does](https://github.com/satijalab/seurat-wrappers/blob/master/DESCRIPTION#L36). We explicitly require Seurat v3.0.1 or higher with SeuratWrappers. If you try to install SeuratWrappers with Seurat v2.X installed, R will automatically install the latest CRAN version of Seurat (or GitHub version, which is synced with the CRAN version, if you installed from GitHub) and overwrite your Seurat v2.X installation. If you then go back and install Seurat v2.X, I don't believe SeuratWrappers will load as we import a lot of Seurat v3.X-specific code.

File: R/cellbrowser.R (hunk 1 above)
Comment (mojaveazure) [reply R/cellbrowser.R:200]: Were you getting error messages when using this function from within SeuratWrappers? If so, then there's a bug in SeuratWrappers somewhere as this package depends on Seurat

File: R/cellbrowser.R (hunk 4 above)
Comment (mojaveazure) [reply R/cellbrowser.R]: The dimensional reduction is a specific class defined by Seurat [here](https://github.com/satijalab/seurat/blob/master/R/objects.R#L136-L149). The equivalent Python code would be

```python
class DimReduc(object):

    def __init__(self, cell_embeddings, feature_loadings, feature_loadings_projected, assay_used, is_global, stdev, key, jackstraw, misc):
        self.cell_embeddings = cell_embeddings
        self.feature_loadings = feature_loadings
        self.feature_loadings_projected = feature_loadings_projected
        self.assay_used = assay_used
        self.is_global = is_global
        self.stdev = stdev
        self.key = key
        self.jackstraw = jackstraw
        self.misc = misc
```

The cell embeddings are just one matrix in this class

```R
# In R
Embeddings(reduction)
```

```python
# In Python
reduction.cell_embeddings
```

Typically, the reductions are named after the technique run (eg. "pca", "umap", "tsne") and we use this to specify which specific cell embeddings we want

```R
# In R
reduction <- object[["pca"]]
embeddings <- Embeddings(reduction)
# or
embeddings <- Embeddings(object[["pca"]])
```
So it's not wrong to have a parameter named `reductions` to refer to dimensional reductions from which we will pull cell embeddings. For example, [`DimPlot`](https://rdrr.io/cran/Seurat/man/DimPlot.html) takes a parameter `reduction` that takes the name of a dimensional reduction. We use this name to pull cell embedding for visualization later. As such, we accept and encourage the use of the term `reductions` to refer to dimensional reductions that will be used later.

I'm not saying you're doing anything wrong per-say, I'm just trying to explain that we don't prefer the term "embeddings" to refer to a dimensional reduction object from which data will later be pulled

**Comment by pcm32** [comment]
Hi! Is this too far from being finalised? This is stopping our https://github.com/ebi-gene-expression-group/r-seurat-scripts/pull/29 to move to Seurat 4 as we depend on the Export to CellBrowser functionality. If that specific functionality is ready for merging, but this PR is going to take longer, I could take that script and add it in the meantime to our repo until this is ready. Thanks!

File: R/cellbrowser.R (hunk 3 above)
Comment (maximilianh) [reply R/cellbrowser.R]: OK, Thanks for the explanation. I'd still really prefer to keep this particular function backwards-compatible, as it's just a few lines and this allows us to use this particular function across the three different Seurat versions. The version problems with Seurat are getting worse and worse and this function isolates us from the Seurat versions.

File: R/cellbrowser.R (hunk 1 above)
Comment (maximilianh) [reply R/cellbrowser.R:200]: We use this function in a little script where the incoming data is a Seurat object. It uses whatever R and Seurat the user has installed and the function is run to convert the data. So it's just another check, it's rare, but some users have different R environments and don't fully know which one they're running the script in, so I'd rather leave the check in.

File: R/cellbrowser.R (hunk 12 above)
Comment (maximilianh) [reply R/cellbrowser.R:410]: Let me know if you think I should use the Misc() function instead of object@misc, but object@misc seems good to me. I'll resolve the question now. Feel free to re-open.

**Comment by maximilianh** [comment]
Sorry for the delay. I was waiting for a few replies and maybe @mojaveazure was too. I've resolved all comments now. Hey @mojaveazure, is this ready to merge? Many thanks for your Sep 2 comment, I didn't realize that Seurat2 and 3 could be installed a the same time. My if-then-elses are still useful, at least for us, because they mean that we can use this function - just the function, not the entire package - with any Seurat version and in this way, the function in Seurat-wrappers can be used both from seurat-wrappers and also from our own data exporters, which work with seurat3 and seurat2, using the same source code.

**Comment by maximilianh** [comment]
I've resolved two merge conflicts, by adding all imports that have been added in the meantime. I hope this is fine, @mojaveazure ?

**Comment by maximilianh** [comment]
@pcm32 I think you can add this to your own repo as a temporary fix, but I hope that Paul can merge it soon. I don't think any of the comments were requests to change something, mostly just suggestions or explanations. Sorry that this somehow fell of the plate...

**Comment by mojaveazure** [comment]
Hi Max, I still see one merge conflict left over on lines 49 to 92 (affects `writeSparseTsvChunks` and `writeSparseMatrix`). Can you fix this conflict? Other than that, this should be good to go

File: R/cellbrowser.R (hunk 8 above)
Comment (maximilianh) [reply R/cellbrowser.R:112]: This runs a few GetAssay calls to find the "best" default matrix for display, in this order: raw counts, scaled or data and checks the value.

File: R/cellbrowser.R (hunk 9 above)
Comment (maximilianh) [reply R/cellbrowser.R:259]: Our development version now has an assay switch, so resolving this.

**Comment by maximilianh** [comment]
Hi @mojaveazure right below your last comment, Github shows: "This branch has no conflicts with the base branch". Are you sure that there is a merge conflict with the base branch, in this particular PR?

**Comment by maximilianh** [comment]
OK, the automated test also ran successfully. 

In the future, I should add an automated test for pbmc3k.final or ideally a much smaller dataset (pmbcSmall?), but I don't know where/how to add such a test...

**Comment by maximilianh** [comment]
Hi @mojaveazure, this looks ready to merge... just checking in if you think there is something for me left to do?

**Comment by mojaveazure** [comment]
Hi Max, sorry for the delay. This looks excellent; thanks for taking care of it!

**Comment by maximilianh** [comment]
Wow, great, thanks Paul!
